### PR TITLE
net: buf: remove data from end of buffers

### PR DESCRIPTION
**Pull request by JordanYates**
Adds a family of functions that remove data from the end of a `struct net_buf`.
This rounds out the capabilities of `struct net_buf`, data can now be added and removed from both the start and end.

Implements #31069

### CODE REVIEW

**Review by tbursztyka — COMMENTED** [review]
lgtm

however, you did not tell why you would need these functions? 
Any use case in mind?

**Comment by JordanYates** [comment]
> Any use case in mind?

Sometimes I wish to attach a temporary piece of context information to a net_buf when passing them between threads via FIFO's (add the context before `net_buf_put` remove after `net_buf_get`).
Context information can be added in 3 places:
1. In the user_data parameter
2. Before the main data (after reserving headroom)
3. After the main data

I have already increased the user_data size and do not wish to increase it further to accommodate occasional requirements.
I already have a defined use for the headroom and don't want to add a secondary use.
Sometimes it just seems cleaner to append temporary information than to prepend it.

I couldn't see a reason why we shouldn't have the option to remove data from the end as well as the front.

**Comment by jukkar** [comment]
> I have already increased the user_data size and do not wish to increase it further to accommodate occasional requirements.
> I already have a defined use for the headroom and don't want to add a secondary use.
> Sometimes it just seems cleaner to append temporary information than to prepend it.
> 
> I couldn't see a reason why we shouldn't have the option to remove data from the end as well as the front.

That is a neat idea. There has been some discussion earlier about this kind of API as normally one would parse, for example networking data, starting from beginning of the buffer, so this kind of API was not really needed. But for this kind of use case, putting stuff at the end of net_buf, this kind of API is indeed useful.

**Comment by jhedberg** [comment]
I'm not opposed to this, but the use case sounds a bit hackish to me, since you're mixing meta-data in the same storage space where the actual data is located. I do understand the reasoning though, since increasing user data would impact all net_buf instances in the system, i.e. the memory overhead would be greater.

**Comment by JordanYates** [comment]
> I'm not opposed to this, but the use case sounds a bit hackish to me, since you're mixing meta-data in the same storage space where the actual data is located. I do understand the reasoning though, since increasing user data would impact all net_buf instances in the system, i.e. the memory overhead would be greater.

I understand that its hacky, but the shakiness of this particular use case doesn't diminish the general usefulness.
If it makes you feel better pretend that I'm using net_buf_simple as a simple stack, push/pop style :)

File: include/net/buf.h [hunk 1]
```diff
@@ -344,7 +344,151 @@ void net_buf_simple_add_le64(struct net_buf_simple *buf, uint64_t val);
 void net_buf_simple_add_be64(struct net_buf_simple *buf, uint64_t val);
 
 /**
- * @brief Push data to the beginning of the buffer.
+ * @brief Remove data from the end of the buffer.
+ *
+ * Removes data from the beginning of the buffer by modifying the buffer length.
```
Comment (jukkar) [inline include/net/buf.h:349]: Aren't we removing data from **the end of the buffer** instead of beginning?

Comment (JordanYates) [reply include/net/buf.h:349]: Yes, I missed this when converting comments. Thanks

**Review by jukkar — APPROVED** [review]
LGTM

**Review by jhedberg — APPROVED** [review]

**Comment by jhedberg** [comment]
Unrelated to this PR, but I'll mention it here anyway: we should really convert the wrapper defines that map net_buf APIs to corresponding net_buf_simple APIs to use static inline functions instead. This would give us better type checks/safety. If anyone wants to submit a PR for that it'd be more than welcome :)

**Comment by JordanYates** [comment]
> If anyone wants to submit a PR for that it'd be more than welcome :)

I'll have a go after this is merged

**Pull request merged**